### PR TITLE
Feature/add psm to objects

### PR DESCRIPTION
**Pull request by msmcconnell**
# PR Details
## Description
This PR updates the motion_computation node to support selection of 1 or more object detection streams which are all synchronized based on the current object detection callbacks. In addition, the PSM message is added as a source of object detection for pedestrians. The method of synchronization is reused based on the approach previously used for mobility path messages. 

To support these changes motion_computation is refactored to have each msg->object conversion written in a separate file. 

## Related Issue
Supports https://github.com/usdot-fhwa-stol/carma-platform/issues/1672

## Motivation and Context
Supports CP use case

## How Has This Been Tested?
Unit testing

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: motion_computation/src/psm_to_external_object_convertor.cpp [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+        tf2::Transform T_m_n_no_heading(R_m_n, tf2::Vector3(0, 0, 0)); // Used to transform the covariance
+
+        // This covariance represents the covariance of the NED frame N-lat,
+        // E-lon, heading- angle east of north This means that the covariance is
+        // in the NED frame, and needs to be transformed to the map frame
+        // clang-format off
+        std::array<double, 36> input_covariance = { 
+          lat_variance, 0, 0,  0, 0, 0,
+          0, lon_variance, 0,  0, 0, 0,
+          0, 0, 0,  0, 0, 0,
```
Comment (msmcconnell) [inline motion_computation/src/psm_to_external_object_convertor.cpp:346]: I probably need a 1 in the third element here to keep the diagonal populated

File: motion_computation/src/psm_to_external_object_convertor.cpp [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+    const std::vector<geometry_msgs::msg::Pose>& poses, double constant_velocity, const rclcpp::Time& start_time,
+    const rclcpp::Duration& step_size, const std::string& frame, double initial_pose_confidence,
+    double initial_vel_confidence) {
+  std::vector<carma_perception_msgs::msg::PredictedState> output;
+  output.reserve(poses.size());
+
+  rclcpp::Time time(start_time);
+
+  for (auto p : poses) {
+    time += step_size;
```
Comment (adev4a) [inline motion_computation/src/psm_to_external_object_convertor.cpp:386]: We're incrementing the time as soon as the loop starts, which means there is no predicted state for 0 time. Which makes sense, but wanted to check if that was the intention

File: motion_computation/src/motion_computation_node.cpp [hunk 3]
```diff
@@ -85,23 +94,36 @@ namespace motion_computation
...
                                                               std::bind(&MotionComputationWorker::predictionLogic, &motion_worker_, std_ph::_1));
-    mobility_path_sub_ = create_subscription<carma_v2x_msgs::msg::MobilityPath>("incoming_mobility_path", 20,
+    
+    mobility_path_sub_ = create_subscription<carma_v2x_msgs::msg::MobilityPath>("incoming_mobility_path", 100,
                                                               std::bind(&MotionComputationWorker::mobilityPathCallback, &motion_worker_, std_ph::_1));
+
+    bsm_sub_ = create_subscription<carma_v2x_msgs::msg::BSM>("incoming_mobility_path", 100,
+                                                              std::bind(&MotionComputationWorker::bsmCallback, &motion_worker_, std_ph::_1));
+
+    psm_sub_ = create_subscription<carma_v2x_msgs::msg::PSM>("incoming_mobility_path", 100,
```
Comment (adev4a) [inline motion_computation/src/motion_computation_node.cpp:104]: fix topic

File: motion_computation/src/motion_computation_node.cpp [hunk 4]
```diff
@@ -85,23 +94,36 @@ namespace motion_computation
     // Setup subscribers
     motion_comp_sub_ = create_subscription<carma_perception_msgs::msg::ExternalObjectList>("external_objects", 1,
                                                               std::bind(&MotionComputationWorker::predictionLogic, &motion_worker_, std_ph::_1));
-    mobility_path_sub_ = create_subscription<carma_v2x_msgs::msg::MobilityPath>("incoming_mobility_path", 20,
+    
+    mobility_path_sub_ = create_subscription<carma_v2x_msgs::msg::MobilityPath>("incoming_mobility_path", 100,
                                                               std::bind(&MotionComputationWorker::mobilityPathCallback, &motion_worker_, std_ph::_1));
+
+    bsm_sub_ = create_subscription<carma_v2x_msgs::msg::BSM>("incoming_mobility_path", 100,
```
Comment (adev4a) [inline motion_computation/src/motion_computation_node.cpp:101]: fix topic

File: motion_computation/src/psm_to_external_object_convertor.cpp (hunk 1 above)
Comment (adev4a) [reply motion_computation/src/psm_to_external_object_convertor.cpp:346]: In that case, do we also need to look into cases where the variance values are 0? There is a case defined where that happens (Line 159)

Comment (msmcconnell) [reply motion_computation/src/psm_to_external_object_convertor.cpp:346]: Hmm that's a fair point. Maybe set them to 1 as well if that happens.

File: motion_computation/src/psm_to_external_object_convertor.cpp [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+    out_msg.confidence = 1.0 - std::min(1.0, fabs(largest_position_std / MAX_POSITION_STD));
+    out_msg.presence_vector |= carma_perception_msgs::msg::ExternalObject::CONFIDENCE_PRESENCE_VECTOR;
+
+  } else if (in_msg.accuracy.presence_vector |
+             carma_v2x_msgs::msg::PositionalAccuracy::ACCURACY_ORIENTATION_AVAILABLE) {
+    // Orientation accuracy available
+
+    lat_variance = 1.0;
+    lon_variance = 1.0;
+    heading_variance = in_msg.accuracy.orientation * in_msg.accuracy.orientation;
```
Comment (msmcconnell) [inline motion_computation/src/psm_to_external_object_convertor.cpp:157]: Add a line here to set out_msg.confidence = 0.1;

File: motion_computation/src/psm_to_external_object_convertor.cpp [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+    // Both accuracies available
+
+      // Compute the position covariance
+    // For computing confidence we will use the largest provided standard
+    // deviation of position
+    largest_position_std = std::max(in_msg.accuracy.semi_major, in_msg.accuracy.semi_minor);
+    lat_variance = in_msg.accuracy.semi_minor * in_msg.accuracy.semi_minor;
+    lon_variance = in_msg.accuracy.semi_major * in_msg.accuracy.semi_major;
+    heading_variance = in_msg.accuracy.orientation * in_msg.accuracy.orientation;
+    position_confidence = 0.1;  // Default will be 10% confidence. If the position accuracy is
```
[the remark this answers is not in the document]
Comment (msmcconnell) [reply motion_computation/src/psm_to_external_object_convertor.cpp:124]: Yeah. looks like out_msg.confidence is used instead but it needed to be set in one more place which I left a comment for

File: motion_computation/src/psm_to_external_object_convertor.cpp [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+        std::array<double, 36> new_cov =
+            tf2::transformCovariance(input_covariance,
+                                     // Per the usage of transformCovariance in tf2_geometry_msgs
+                                     // this frame should be the transform between the map which the pose
+                                     // is in an the target frame.
+                                     T_m_n_no_heading);
+
+        // Populate message
+        geometry_msgs::msg::PoseWithCovariance pose;
+        pose.pose.position.x = T_m_s.getOrigin().getX();
```
[the remark this answers is not in the document]
Comment (msmcconnell) [reply motion_computation/src/psm_to_external_object_convertor.cpp:362]: This is correct. Its a weirdly names function but it means get the translation

File: motion_computation/src/psm_to_external_object_convertor.cpp (hunk 2 above)
Comment (msmcconnell) [reply motion_computation/src/psm_to_external_object_convertor.cpp:386]: Yeah thats intentional. These are just future predictions

**Review by adev4a — APPROVED** [review]

**Pull request merged**